returnn/tensor/dim.py

```diff
@@ -0,0 +1,99 @@
+"""
+Represents a dimension of a tensor.
+A dimension can come with further information such as individual sequence lengths.
+
+This identifies one axis/dimension, like a time-dimension, etc.
+This was called ``DimensionTag`` earlier, and referred to as dimension tag.
+
+This is used by :class:`Tensor` (earlier ``Data``). See :func:`Tensor.dims`.
+This would be passed as ``dims`` when creating a :class:`Tensor` instance.
+
+It is not to specify the specific axis in a specific Tensor,
+but to specify the content and dimension.
+I.e. if we have the same Dim for two Data instances,
+the dimensions should match. I.e.:
+
+    data1.dims[i] == data2.dims[j]
+      =>  data1.raw_tensor.shape[i] == data2.raw_tensor.shape[j]
+
+This also includes further information such as sequence lengths
+or a vocabulary.
+
+Deprecated: We differentiate between the batch dim, spatial dim or feature dim,
+although that is just flag and in many contexts there is no real difference
+between a spatial dim and a feature dim (the batch dim is often handled differently).
+
+"""
+
+
+from __future__ import annotations
+from typing import Optional, Union
+
+from ._dim_extra import _DimExtra, _DimMixin, DimTypes
+from . import tensor as _t
+
+
+class Dim(_DimMixin):
+    """
+    Represents a dimension of a tensor.
+    This potentially comes with further information such as individual sequence lengths.
+    See the module docstring.
+    """
+
+    Types = DimTypes  # old alias
+
+    __slots__ = ("name", "capacity", "size", "dyn_size_ext", "_extra")
+
+    name: Optional[str]
+    capacity: Optional[int]  # shape[axis] in the raw tensor (might need power-of-two or static shape), None if dynamic
+    size: Optional[int]  # shape[axis] in the represented tensor if static, None if dynamic, then dyn_size_ext
+    dyn_size_ext: Optional[_t.Tensor]
+    _extra: Optional[_DimExtra]
+
+    def __init__(
+        self,
+        dimension: Optional[Union[int, _t.Tensor]],
+        *,
+        name: Optional[str] = None,
+        capacity: Optional[int] = None,
+        dyn_size_ext: Optional[_t.Tensor] = None,
+        description: Optional[str] = None,
+        **kwargs,
+    ):
+        self.name = name or description
```

Comment:
Right now if `name` is passed, `description` is ignored. I think `description` could be a useful addition for debugging purposes as an attribute separate to `name`. @albertz what do you think?

Reply:
Currently `description` in all old code (also external code) is treated like a `name`.

Reply:
And in this PR, `description` and `name` are just aliases of each other.